### PR TITLE
add doc on performance principles re: rules and files to consider

### PR DESCRIPTION
**Pull request by khorne3**
# Thanks for improving Semgrep Docs 😀

### Please ensure

- [x] A subject matter expert (SME) reviews the content
- [x] A technical writer reviews the content or PR
- [x] This change has no security implications or else you have pinged the security team

### CODE REVIEW

File: docs/kb/rules/rule-file-perf-principles.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+You can benchmark your rules by adding the `--time` flag to your `semgrep scan`
+command. When you use this flag, your results return with a timing summary; if
+your output format is JSON, you'll see times for each rule-target pair.
+
+## Files
+
+Generally, the time required to scan files scales linearly with the number of
+files scanned, but file size is still important. Overall, the time taken is
+**time for setup work + time for matching**. For setup work, files aren’t
+analyzed alone but in groups of mutually dependent files (SCCs).
```
Comment (khorne3) [inline docs/kb/rules/rule-file-perf-principles.md:35]: @emjin what does SCC stand for?

Comment (fuzzmaestro1) [reply docs/kb/rules/rule-file-perf-principles.md:35]: SCC stands for  strongly connected components, yes @emjin

**Review by fuzzmaestro1 — APPROVED** [review]

File: docs/kb/rules/rule-file-perf-principles.md (hunk 1 above)
Comment (emjin) [reply docs/kb/rules/rule-file-perf-principles.md:35]: That's correct!

File: docs/kb/rules/rule-file-perf-principles.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+---
+
+import MoreHelp from "/src/components/MoreHelp"
+
+# Performance principles for rules and files to abide by when scanning repositories
+
+## Rules
+
+The amount of time required for rules to run scales better than linearly when
+adding interfile rules, which are those with `interfile: true` in the options.
```
Comment (s-santillan) [inline docs/kb/rules/rule-file-perf-principles.md:16]:
```suggestion
adding interfile rules, which are those with `interfile: true` in the `options` key.
```

File: docs/kb/rules/rule-file-perf-principles.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+That is, doubling the number of interfile rules increases the runtime, but not
+by double. However, some rules run faster than others, and adding a slow rule
+when all the rest are fast can cause a significant slowdown.
+
+Rules are slower if the sub-patterns, such as `pattern: <... $X ...>`, result in
+a greater number of matches. When writing rules, pay special attention to the
+problems raised by sub-pattern matches. The most important factor for runtime is
+the time spent adding to various lists/sets.
```
Comment (s-santillan) [inline docs/kb/rules/rule-file-perf-principles.md:24]:
```suggestion
the time spent adding to various lists or sets.
```

**Review by s-santillan — APPROVED** [review]

**Comment by s-santillan** [comment]
It's the first time I've heard of strongly-connected components! Is this a thing in our tree-sitter / engine?

**Pull request merged**